File: internal/tui/selector/selector.go
package selector

import (
	"fmt"
	"strings"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
	"github.com/whoisjordangarcia/dotfiles/internal/installer"
	"github.com/whoisjordangarcia/dotfiles/internal/tui/theme"
)

const logo = `  ▄▄█▀▀██▄   ██            ▄██
▄██▀    ▀██▄ ██             ██
██▀      ▀███████  ▄██▀██▄  ██▄████▄ ▀███  ▀███ ▀████████▄
██        ██ ██   ██▀   ▀██ ██    ▀██  ██    ██   ██    ██
██▄      ▄██ ██   ██     ██ ██     ██  ██    ██   ██    ██
▀██▄    ▄██▀ ██   ██▄   ▄██ ██▄   ▄██  ██    ██   ██    ██
  ▀▀████▀▀   ▀████ ▀█████▀  █▀█████▀   ▀████▀███▄████  ████▄`

// Result holds the user's module selection.
type Result struct {
	Components []installer.Component
	Selected   []bool
	Confirmed  bool
}

// Model is the Bubble Tea model for the module selector screen.
type Model struct {
	components []installer.Component
	selected   []bool
	cursor     int
	system     string
	done       bool
	quitting   bool
	width      int
	height     int
}

// New creates a new module selector model.
func New(system string, components []installer.Component) Model {
	selected := make([]bool, len(components))
	for i := range selected {
		selected[i] = true
	}
	return Model{
		components: components,
		selected:   selected,
		system:     system,
	}
}

// GetResult returns the selection result after the model is done.
func (m Model) GetResult() Result {
	return Result{
		Components: m.components,
		Selected:   m.selected,
		Confirmed:  m.done && !m.quitting,
	}
}

func (m Model) Init() tea.Cmd {
	return nil
}

func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	switch msg := msg.(type) {
	case tea.WindowSizeMsg:
		m.width = msg.Width
		m.height = msg.Height
	case tea.KeyMsg:
		switch msg.String() {
		case "ctrl+c", "q":
			m.quitting = true
			return m, tea.Quit
		case "up", "k":
			if m.cursor > 0 {
				m.cursor--
			}
		case "down", "j":
			if m.cursor < len(m.components)-1 {
				m.cursor++
			}
		case " ", "x":
			m.selected[m.cursor] = !m.selected[m.cursor]
		case "a":
			for i := range m.selected {
				m.selected[i] = true
			}
		case "n":
			for i := range m.selected {
				m.selected[i] = false
			}
		case "enter":
			m.done = true
			return m, tea.Quit
		}
	}
	return m, nil
}

func (m Model) View() string {
	if m.done || m.quitting {
		return ""
	}

	var b strings.Builder

	// Logo
	b.WriteString(theme.Logo.Render(logo))
	b.WriteString("\n\n")

	// Header
	selectedCount := 0
	for _, s := range m.selected {
		if s {
			selectedCount++
		}
	}

	header := theme.Title.Render("  Module Selection") + "  " +
		theme.Badge.Render(fmt.Sprintf("[%d/%d]", selectedCount, len(m.components))) + "  " +
		theme.Subtitle.Render(m.system)
	b.WriteString(header)
	b.WriteString("\n")
	b.WriteString(theme.HelpSep.Render("  " + strings.Repeat("─", 50)))
	b.WriteString("\n\n")

	// Module list
	for i, comp := range m.components {
		var line string

		if m.cursor == i {
			cursor := theme.CursorStyle.Render("❯ ")
			checkbox := theme.Unselected.Render("○")
			if m.selected[i] {
				checkbox = theme.Selected.Render("●")
			}
			name := theme.ActiveItem.Render(comp.Name)
			line = fmt.Sprintf("  %s%s  %s", cursor, checkbox, name)
		} else {
			checkbox := theme.Unselected.Render("○")
			if m.selected[i] {
				checkbox = theme.Selected.Render("●")
			}
			name := comp.Name
			if m.selected[i] {
				name = lipgloss.NewStyle().Foreground(theme.White).Render(name)
			} else {
				name = theme.Unselected.Render(name)
			}
			line = fmt.Sprintf("    %s  %s", checkbox, name)
		}

		b.WriteString(line + "\n")
	}

	// Help bar
	b.WriteString("\n")
	b.WriteString(theme.HelpSep.Render("  " + strings.Repeat("─", 50)))
	b.WriteString("\n")

	helpItems := []struct{ key, desc string }{
		{"space", "toggle"},
		{"a", "all"},
		{"n", "none"},
		{"↵", "install"},
		{"q", "quit"},
	}
	var helpParts []string
	for _, h := range helpItems {
		helpParts = append(helpParts,
			theme.HelpKey.Render(h.key)+theme.HelpValue.Render(" "+h.desc))
	}
	help := "  " + strings.Join(helpParts, theme.HelpSep.Render(" · "))
	b.WriteString(help)

	return b.String()
}

File: crates/otobun/src/tui/selector.rs
use crate::installer::components::Component;
use crate::tui::theme;
use crossterm::event::{self, Event, KeyCode, KeyEventKind};
use ratatui::text::{Line, Span};
use ratatui::widgets::Paragraph;
use ratatui::Frame;

pub struct SelectorResult {
    pub components: Vec<Component>,
    pub selected: Vec<bool>,
    pub confirmed: bool,
}

pub struct Selector {
    components: Vec<Component>,
    selected: Vec<bool>,
    cursor: usize,
    system: String,
    done: bool,
    confirmed: bool,
}

impl Selector {
    pub fn new(components: Vec<Component>, system: String) -> Self {
        let selected = vec![true; components.len()];
        Self {
            components,
            selected,
            cursor: 0,
            system,
            done: false,
            confirmed: false,
        }
    }

    pub fn run(mut self, terminal: &mut ratatui::DefaultTerminal) -> anyhow::Result<SelectorResult> {
        while !self.done {
            terminal.draw(|frame| self.draw(frame))?;
            if let Event::Key(key) = event::read()? {
                if key.kind == KeyEventKind::Press {
                    self.handle_key(key.code);
                }
            }
        }
        Ok(SelectorResult {
            components: self.components,
            selected: self.selected,
            confirmed: self.confirmed,
        })
    }

    fn handle_key(&mut self, key: KeyCode) {
        match key {
            KeyCode::Char('j') | KeyCode::Down => {
                if self.cursor < self.components.len().saturating_sub(1) {
                    self.cursor += 1;
                }
            }
            KeyCode::Char('k') | KeyCode::Up => {
                if self.cursor > 0 {
                    self.cursor -= 1;
                }
            }
            KeyCode::Char(' ') | KeyCode::Char('x') => {
                self.selected[self.cursor] = !self.selected[self.cursor];
            }
            KeyCode::Char('a') => {
                self.selected.fill(true);
            }
            KeyCode::Char('n') => {
                self.selected.fill(false);
            }
            KeyCode::Enter => {
                self.confirmed = true;
                self.done = true;
            }
            KeyCode::Char('q') | KeyCode::Esc => {
                self.confirmed = false;
                self.done = true;
            }
            _ => {}
        }
    }

    fn draw(&self, frame: &mut Frame) {
        let area = frame.area();

        let selected_count = self.selected.iter().filter(|s| **s).count();
        let total = self.components.len();

        let mut lines = Vec::new();

        // Logo
        lines.push(Line::styled("  otobun", theme::logo()));
        lines.push(Line::raw(""));

        // Title with badge
        lines.push(Line::from(vec![
            Span::styled("  Module Selection ", theme::title()),
            Span::styled(format!("[{selected_count}/{total}] "), theme::badge()),
            Span::styled(&self.system, theme::subtitle()),
        ]));
        lines.push(Line::styled(
            "  ──────────────────────────────────────────────────",
            theme::help_sep(),
        ));
        lines.push(Line::raw(""));

        // Component list
        for (i, comp) in self.components.iter().enumerate() {
            let is_cursor = i == self.cursor;
            let is_selected = self.selected[i];

            let cursor_str = if is_cursor { "  ❯ " } else { "    " };
            let check = if is_selected { "●" } else { "○" };

            let check_style = if is_selected {
                theme::selected()
            } else {
                theme::unselected()
            };
            let name_style = if is_cursor {
                theme::active_item()
            } else if is_selected {
                ratatui::style::Style::default().fg(theme::WHITE)
            } else {
                theme::unselected()
            };

            lines.push(Line::from(vec![
                Span::styled(cursor_str, theme::cursor()),
                Span::styled(check, check_style),
                Span::raw(" "),
                Span::styled(&comp.name, name_style),
            ]));
        }

        // Footer
        lines.push(Line::raw(""));
        lines.push(Line::styled(
            "  ──────────────────────────────────────────────────",
            theme::help_sep(),
        ));
        lines.push(Line::from(vec![
            Span::raw("  "),
            Span::styled("space", theme::help_key()),
            Span::styled(": ", theme::help_sep()),
            Span::styled("toggle", theme::help_value()),
            Span::styled(" · ", theme::help_sep()),
            Span::styled("a", theme::help_key()),
            Span::styled(": ", theme::help_sep()),
            Span::styled("all", theme::help_value()),
            Span::styled(" · ", theme::help_sep()),
            Span::styled("n", theme::help_key()),
            Span::styled(": ", theme::help_sep()),
            Span::styled("none", theme::help_value()),
            Span::styled(" · ", theme::help_sep()),
            Span::styled("↵", theme::help_key()),
            Span::styled(": ", theme::help_sep()),
            Span::styled("install", theme::help_value()),
            Span::styled(" · ", theme::help_sep()),
            Span::styled("q", theme::help_key()),
            Span::styled(": ", theme::help_sep()),
            Span::styled("quit", theme::help_value()),
        ]));

        let paragraph = Paragraph::new(lines);
        frame.render_widget(paragraph, area);
    }
}
